<p align="center">
<img src='../../img/VerticaMLPython.png' width="180px">
</p>

# Vertica ML Python Exercise 3

During this exercice, we will:
<ul>
    <li> Draw correlation matrix
    <li> Understand the difference between dependance and correlation
    <li> Understand how to use the correct correlation method
</ul>
## Initialization

Let's create a cursor using the vertica_cursor function

In [1]:
from vertica_ml_python.utilities import vertica_cursor
cur = vertica_cursor("VerticaDSN")

During this study, we will work with the churn dataset. You'll need to read the 'churn.csv' file.

In [2]:
from vertica_ml_python.utilities import read_csv
from vertica_ml_python.utilities import drop_table
drop_table('churn', cur)
churn = read_csv('churn.csv', cur)
print(churn)

The table churn was successfully dropped.
The table public.churn has been successfully created.


<object>  Name: churn, Number of rows: 7043, Number of columns: 21


This dataset contains many information of 7043 customers including:
<ul>
    <li><b>customerID: </b>Customer ID
    <li><b>gender: </b>Whether the customer is a male or a female
    <li><b>SeniorCitizen: </b>Whether the customer is a senior citizen or not (1, 0)
    <li><b>Partner: </b>Whether the customer has a partner or not (Yes, No)
    <li><b>Dependents: </b>Whether the customer has dependents or not (Yes, No)
    <li><b>tenure: </b>Number of months the customer has stayed with the company
    <li><b>PhoneService: </b>Whether the customer has a phone service or not (Yes, No)
    <li><b>MultipleLines: </b>Whether the customer has multiple lines or not (Yes, No, No phone service)
    <li><b>InternetService: </b>Customer’s internet service provider (DSL, Fiber optic, No)
    <li><b>OnlineSecurity: </b>Whether the customer has online security or not (Yes, No, No internet service)
    <li><b>OnlineBackup: </b>Whether the customer has online backup or not (Yes, No, No internet service)
    <li><b>DeviceProtection: </b>Whether the customer has device protection or not (Yes, No, No internet service)
    <li><b>TechSupport: </b>Whether the customer has tech support or not (Yes, No, No internet service)
    <li><b>StreamingTV: </b>Whether the customer has streaming TV or not (Yes, No, No internet service)
    <li><b>StreamingMovies: </b>Whether the customer has streaming movies or not (Yes, No, No internet service)
    <li><b>Contract: </b>The contract term of the customer (Month-to-month, One year, Two year)
    <li><b>PaperlessBilling: </b>Whether the customer has paperless billing or not (Yes, No)
    <li><b>PaymentMethod: </b>The customer’s payment method (Electronic check, Mailed check, Bank transfer (automatic), Credit card (automatic))
    <li><b>MonthlyCharges: </b>The amount charged to the customer monthly
    <li><b>TotalCharges: </b>The total amount charged to the customer
    <li><b>Churn: </b>Whether the customer churned or not (Yes or No)
</ul>
The purpose is to understand the correlations and dependances between all these variables.

## Correlations and Dependances

Let's explore the data by displaying descriptive statistics of all the columns.

In [3]:
churn.describe(method = "categorical")

<object>

Many features are categorical and need to be encoded first. However, it is not needed during this exercice as the main purpose is to understand the different type of correlations. 

<b>Question 1: </b>The linear correlation is computed using the 'pearson' method. Draw the Correlation Matrix using the Pearson coefficient. Explain the linear correlations and dependences between the different features.

<b>Question 2: </b>There is no perfect way to deal with non-linear correlations between the different features. Spearman coefficients will identify Monotonic relationships between variables. Draw the Correlation Matrix using the Spearman coefficient. Explain the monotonic relationships between the different features.

<b>Question 3: </b>Kendall coefficients can be very expensive to compute but they will bring very useful information. They can explain complex relationship between the different features. Draw the Correlation Matrix using the Kendall coefficient. Explain the relationships between the different features.

<b>Question 4: </b>The point Biserial method explains the link between a binary variable and a numerical feature. It can be used only if one of the feature is binary, otherwise it will return 0. Draw the Correlation Matrix using the the point Biserial. In this use-case this method is not very efficient, however do not forget that it is always a possibility.

<b>Question 5: </b>The last technique is the Cramer's V which can explain link between categorical columns. It will range only between 0 and 1. This technique is very important as it helps to understand the link between the categorical features which can represent a huge part of the dataset. Draw the Correlation Matrix using the Cramer's V coefficient. Explain the relationships between the different features.

<b>Question 6: </b>Explain which features will be useful to predict churn and what type of ML algorithms you will probably not use for the prediction.